worlds/osrs/__init__.py

```diff
@@ -0,0 +1,616 @@
+            can_spin = state.can_reach(RegionNames.Sheep, None, self.player) and state.can_reach(
+                RegionNames.Spinning_Wheel, None, self.player)
+            can_pot = state.can_reach(RegionNames.Clay_Rock, None, self.player) and state.can_reach(
+                RegionNames.Barbarian_Village, None, self.player)
+            can_tan = state.can_reach(RegionNames.Milk, None, self.player) and state.can_reach(
+                RegionNames.Al_Kharid, None, self.player)
+
+            mould_access = state.can_reach(RegionNames.Al_Kharid, None, self.player) or state.can_reach(
+                RegionNames.Rimmington, None, self.player)
+            if self.allow_brutal_grinds:
+                # Only force killing barbarians for moulds in brutal grinds
+                mould_access = mould_access or state.can_reach(RegionNames.Barbarian_Village, None, self.player)
+            can_silver = state.can_reach(RegionNames.Silver_Rock, None, self.player) and state.can_reach(
+                RegionNames.Furnace, None, self.player) and mould_access
+            can_gold = state.can_reach(RegionNames.Gold_Rock, None, self.player) and \
+                       state.can_reach(RegionNames.Furnace, None, self.player) and mould_access
+
+            can_train = can_spin or can_pot or can_tan
+            if not self.allow_brutal_grinds:
+                if level > 5:
+                    can_tran = can_pot or can_tan or can_gold
+                if level > 16:
+                    can_train = can_tan or can_gold or can_silver
+            return can_train
+        if skill == "Cooking":
+            # Meat and Chicken can be found with milk and eggs.
+            can_bread = state.can_reach(RegionNames.Wheat, None, self.player) and state.can_reach(
+                RegionNames.Windmill, None, self.player)
+            can_train = state.can_reach(RegionNames.Milk, None, self.player) or \
+                        state.can_reach(RegionNames.Egg, None, self.player) or state.can_reach(
+                RegionNames.Shrimp, None, self.player) or can_bread
+            if not self.allow_brutal_grinds:
+                if level > 15:
+                    can_train = can_train and state.can_reach(RegionNames.Fly_Fish, None,
+                                                              self.player) and self.can_reach_skill(state,
+                                                                                                    "fishing", 20)
+            return can_train
+        if skill == "Runecraft":
+            return state.has(ItemNames.QP_Rune_Mysteries, self.player)
+        if skill == "Magic":
+            return state.can_reach(RegionNames.Mind_Runes, None, self.player)
+        # If it's not listed here, it can be trained just fine without needing locations
+        # print(f"Attempting to check for reaching level {level} in {skill} which does not have rules set so it's fine")
+        return True
```

Comment:
Just to drive this point home, this is quite terrible.

Consider a location with no access requirements. This location's access rule might be checked hundreds if not thousands of times during a multiworld generation.

Instead of setting its access rule to `lambda state: True`, which means it gets to instantly evaluate as "True" every time it's checked, instead you force it to check itself against *every skill in the game* first. *Every time the access rule is checked, thousands of times*.